simuleval/agents/pipeline.py

```diff
@@ -88,3 +89,170 @@ def __repr__(self) -> str:
 
     def __str__(self) -> str:
         return self.__repr__()
+
+
+class TreeAgentPipeline(AgentPipeline):
+    """A pipeline which passes intermediate outputs in a directed acyclic graph
+    Note: the target_type will be a "_"-concatenated list of types. If the
+        argument --output-index is specified, `pop` will return the result
+        corresponding to this index.
+
+    Examples:
+        ```
+        pipeline = {
+            ClassOne: [ClassTwo],
+            ClassTwo: [ClassThree],
+            ClassThree: []
+        }
+        ```
+        In this case, ClassOne is the root input node (no parent nodes)
+            and ClassThree is the leaf output node (no child nodes)
+        ```
+        pipeline = {
+            ClassOne: [ClassTwo, ClassThree],
+            ClassTwo: [],
+            ClassThree: [],
+        }
+        ```
+        In this case, ClassOne is the input node,
+            and ClassTwo, ClassThree are output nodes.
+
+        Note: currently does not support multiple inputs to a node or
+            multiple root nodes
+
+    Attributes:
+        pipeline (dict): a dict mapping agent class to the downstream class(es).
+            pipeline is only used to initialize the class using `from_args`
+    """
+
+    pipeline: dict = {}
+
+    def __init__(
+        self,
+        module_dict: Dict[GenericAgent, List[Type[GenericAgent]]],
+        args,
+    ) -> None:
+        (
+            self.source_module,
+            self.target_modules,
+            self.module_dict,
+        ) = TreeAgentPipeline.check_pipeline_types(module_dict)
+
+        self.output_index = args.output_index
+        if self.output_index is not None:
+            assert len(self.target_modules) > args.output_index
+
+    @classmethod
+    def add_args(cls, parser: ArgumentParser):
+        super().add_args(parser)
+        parser.add_argument(
+            "--output-index",
+            type=int,
+            default=None,
+            help="If specified, `pop` will only output the result at this index.",
+        )
+
+    @classmethod
+    def get_instance_from_class(cls, klass, module_dict):
+        ins_list = [ins for ins in module_dict.keys() if isinstance(ins, klass)]
+        assert len(ins_list) == 1, f"Instances of {klass}: {ins_list}"
+        return ins_list[0]
```

Comment:
Is the assumption here that the classes wont be repeated in the pipeline?
How about a pipeline where we want ASR and translated text output,  there can be two different instances of  SPM detokenizer agent right?